### DISCUSSION TITLE
Avoiding disposing when StateNotifierProvider gets invalidated

### DISCUSSION THREAD

**Discussion by andreasgangso** [rrousselGit/riverpod#2325] [General]
Use case: i want to have StateNotifiers which are not controlled by riverpod, but I want to use riverpod's StateNotifierProvider for easy state access.
However, this isn't easy because riverpod is disposing the notifiers on [this line](https://github.com/rrousselGit/riverpod/blob/c18e93e53d62be5f1ae1db255ad365e1ef0a2a71/packages/riverpod/lib/src/state_notifier_provider/base.dart#L190) which is run [when the provider is invalidated](https://github.com/rrousselGit/riverpod/blob/c18e93e53d62be5f1ae1db255ad365e1ef0a2a71/packages/riverpod/lib/src/framework/element.dart#L301).

To help such use cases, I'm wondering if it's a good idea to submit a PR which gives StateNotifierProvider() an argument to prevent the notifier from being disposed. Alternatively, I could override dispose() on the StateNotifier.
But I'm worried that future StateNotifierProvider changes could mess up the StateNotifier in other ways.
Anyone got any thoughts on this?

Here's concrete code showing my setup.
- The player state notifier (a StateNotifier which is not supposed to be disposed by riverpod). [player_state_notifier.dart](https://github.com/bcc-code/brunstadtv-app/blob/master/external_projects/bccm_player/lib/src/state/player_state_notifier.dart).
- The "plugin" state notifier (a StateNotifier which creates and maintains a list of PlayerStateNotifiers): [plugin_state_notifier.dart](https://github.com/bcc-code/brunstadtv-app/blob/master/external_projects/bccm_player/lib/src/state/plugin_state_notifier.dart).
- The various StateNotifierProviders: [player_provider.dart](https://github.com/bcc-code/brunstadtv-app/blob/master/external_projects/bccm_player/lib/src/plugins/riverpod/providers/player_provider.dart).

**Comment by andreasgangso** [comment]
@rrousselGit I ended up overriding dispose() in my package: [player_state_notifier.dart](https://github.com/bcc-code/bccm-player/blob/3f015cdf1623b99efdef0bf0408c9ce5da22037f/lib/src/state/player_state_notifier.dart#L24)
But again, worried this is a bit hacky.

Would creating a PR as I described be useful/accepted? I don't want to waste either your time or mine.

**Reply by rrousselGit** [reply]
Hello! No. This is a sign of a bad practice, and to begin with, StateNotifierProvider is bound to disappear. Notifier/AsyncNotifier are the way to go now.

**Reply by andreasgangso** [reply]
The _provider_ is bound to disappear, but I'm talking about the underlying StateNotifier, which doesn't necessarily have to disappear together with the provider.
Some of my StateNotifiers are supposed to be persistent and usable directly, with riverpod just being a convenient way to watch the state.
